src/NServiceBus.Core/Hosting/Helpers/AssemblyScanner.cs

```diff
@@ -141,7 +144,12 @@ bool TryLoadScannableAssembly(string assemblyPath, AssemblyScannerResults result
 
             try
             {
+#if NETCOREAPP2_1
+                var context = AssemblyLoadContext.GetLoadContext(Assembly.GetExecutingAssembly());
```

Comment:
at least for functions this seems the way to go to ensure that everything is loaded into the function execution specific context. I'm mostly wondering whether there is a case where this behavior would be undesired and this behavior would be breaking (so it should be made opt-in instead), but I can't really come up with a specific scenario though.

Reply:
>  If NServiceBus is loaded into a non-default AssemblyLoadContext, we need to load all scanned type into the same context, or things break, which is the entire reason for this change.

according to my tests, this isn't a major problem. At least in Azure Functions, it doesn't seem to matter into what assembly load context you're loading the assemblies, the type comparison works as expected even across different load contexts. Of course, you want all the endpoint related assemblies to load into the function-specific context because it's purpose is to avoid  version conflicts and that indeed can be a source of problems.

Reply:
> according to my tests, this isn't a major problem

Then you must not have been testing the right thing. The entire reason this PR exists in the first place is that there is a problem with Azure Functions, and Sean and I discovered that it is exactly because of types being loaded into different load contexts.

If types are loaded into different load contexts, type comparisons break because you'll end up with the assembly scanner loading the same type into two different contexts.

Reply:
I've been testing every possible way I could come up with and I could not reproduce type comparison check problems when using different contexts (the repro I got from Sean had issues because it wouldn't load the assembly at all). If you have a repro, please let me know, it would be great to test it with our changes. I'm not sure whether there is special behavior in the functions assembly load context (which is a custom context implementation) which does special tricks in regards to using the default load context as backup or whether that is the default behavior with AssemblyLoadContexts though.

Reply:
I have put together a repro that demonstrates the problem: https://github.com/bording/LoadContextRepro

You can see that types loaded in a custom `AssemblyLoadContext` are not equal to the same type loaded into the default context.

You can also see an example of what happens in the assembly scanner when Azure Functions would invoke it through a custom context. 

If you have any more questions about this, send me a message in slack so we can set up a call to discuss it.

Reply:
thanks for the repro @bording. I had quite a hard time figuring out why I can't reproduce the same behavior in my slightly different repo. In the end, at least when using functions, it looks like it boils down to whether the loaded assemblies are located. In your repro, all the assemblies are coming from entirely different locations, none of them within the probing paths. However, as soon as you move the assembly with the shared interface (Loader.dll in your case) into the scanning path (e.g. by referencing it, which is what the functions scenario currently assumes), the behavior changes a little bit to the following:

```
            var assembly = Assembly.LoadFrom(Path.GetFullPath(assemblyPath));

            var handlerType = assembly.GetTypes().Single(t => t.Name == "TestHandler");
            var handlerContext = AssemblyLoadContext.GetLoadContext(handlerType.Assembly); // this is the default app context, as expected
            var interfaceType = handlerType.GetInterface("IHandler");
            var loadContext = AssemblyLoadContext.GetLoadContext(interfaceType.Assembly); // but this is the function app context

             var equal = interfaceType == typeof(IHandler); //true
```

Let's chat about this later today.